### ISSUE TITLE
State of queues and streams

### ISSUE THREAD

**Issue by jpsamaroo** [JuliaGPU/AMDGPU.jl#337]
At this moment, AMDGPU has to manage two different kinds of hardware queues: HSA queues, and HIP streams. Kernels written in Julia (via AMDGPU, GPUArrays, or KernelAbstractions) use HSA queues, while kernels written in C++ are launched on HIP streams. When kernels are launched from Julia, we record (within each `ROCArray` argument) the signal associated with the kernel; when we launch kernels from HIP, we record the HIP stream. Then, future operations will synchronize on those signals and streams to ensure that kernels don't simultaneously access the same buffers.

Currently, queues and streams are globally managed, but it is my hope to somehow implement task-local queues/streams. One caveat is that, from testing, HSA appears to only be willing to allocate 128 queues (on my system at least). One possible workaround is that we could manually multiplex queues in AMDGPU (like what KA does today), using the kernel's arguments to determine which of a limited set of queues to use. I haven't checked if HIP's streams also have this hard limit; if they don't, it might be possible to use them to launch work, since they're really just HSA queues under the hood.

An additional observation I've made from testing is that HSA queues may actually execute in-order, even though they were previously thought to execute out-of-order (and the HSA AQL API makes it seems like that should be the case). If this is really true, then we'll definitely want to implement the above task-local queue system to make kernel parallelism easier to attain.

Additionally, while KernelAbstractions currently has an explicit synchronization system, it will be moving to a per-task stream-based system soon. We should be able to make this work by launching barriers just after kernel launches, to block the queue and make it act like a stream. How HIP streams will be handled, though, is not clear to me.

**Comment by luraess** [comment]
Thanks for summarising the state here 🙏 

From a HPC usage perspective, as needed in https://github.com/PTsolvers/FastIce.jl, here are some additional thoughts and experience feedback:

- There seem to be a limit in the max number of queues, and even a "harder" limit on the max number of queues with priorities. Using ROCQueues as analogy to CuStreams thus does not work and results in a HSA failure (OOM error).
- Kernels launched on the same queue seem to execute in order; this is nice as it simplifies sync and one can use queues (up to a given extend to execute out-of-order). This is also less nice as the current situation has a limit on the max number of queues (~7-8 high priority ones on MI250x with [this code](https://github.com/PTsolvers/FastIce.jl/blob/lr/levset/scripts3D_amd/FastIce3D_visc_amd_opt_mpi_perf.jl)) and this situation could become limiting in case many small kernels need to be executed out-of-order.
- For better performance, it seems good practice to pass an already initialised `ROCSignal()` signal, making sure to manually handle the signal store and release as in [this code](https://github.com/PTsolvers/FastIce.jl/blob/lr/levset/scripts3D_amd/FastIce3D_visc_amd_opt_mpi_perf.jl). It seems also to be fine to wait on that signal instead of the signal returned by the `@roc` call. This is a rather duplicated situation; it could be maybe good to have the signal re-use build into the `@roc` API?

As general thoughts (which may be biased by the way things work in CUDA), and given the recent experiences, it would be nice to have AMDGPU to initialise at leat 3 queues, 1 for each prio, or find out how many can be init maximally and split them among priorities. Then the user could just launch kernels on these already initialised queues. Signals could also be initialised upon first call and then re-used, dropping the `ROCKernelSignal` and only using the `ROCSignal` passed to the kernel maybe. This would allow to use queues and signal in a very handy way and remove the burden from the end user to handle these things explicitly. Ultimately, one could think about exposing something like `rocstreams` that would handle queue priorities and signal dispatching under the hood and offer a CUDA analogous API. I guess having API's with similar design between AMDGPU and CUDA could also be beneficial to the end user as it would reduce the barrier of moving from CUDA to AMDGPU.

To be further discussed
